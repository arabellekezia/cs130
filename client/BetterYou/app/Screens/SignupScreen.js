import React from "react";
import { SafeAreaView, StyleSheet, Text, View } from "react-native";
import AppText from "../components/AppText";
import AppTextInput from "../components/AppTextInput";
import TextButton from "../components/TextButton";
import server from "../utils/server";
import { storeUserToken } from "../utils/token";

function SignupScreen() {
  const [name, setName] = React.useState("");
  const [email, setEmail] = React.useState("");
  const [password, setPassword] = React.useState("");
  const [confirmPassword, setConfirmPassword] = React.useState("");
  const [err, setError] = React.useState({
    name: false,
    email: false,
    password: false,
    accountExists: false,
  });

  function signUp() {
    if (!validateFields({ name, email, password, confirmPassword }, setError)) {
      console.log(err);
      return;
    }

    let formdata = new FormData();
    formdata.append("name", name);
    formdata.append("email", email);
    formdata.append("password", password);

    server
      .post("/register", formdata, {
        headers: {
          "Content-Type": "multipart/form-data",
        },
      })
      .then(async (res) => {
        console.log("Successfully signed up!");
        await storeUserToken(res.data);
      })
      .catch((err) => {
        if (err.response.status === 400) {
          setError({ email: true, accountExists: true });
        }
      });
  }

  function displayErrorMessage(err) {
    if (err.name) {
      return <ErrorMessage message="Please enter your full name." />;
    }

    if (err.accountExists) {
      return (
        <ErrorMessage message="An account with this email already exists." />
      );
    }
    
    if (err.email) {
      return <ErrorMessage message="Please use a valid email." />;
    }

    if (err.password) {
      if (password.length < 8) {
        return (
          <ErrorMessage message="Password must be at least 8 characters long!" />
        );
      }
      return <ErrorMessage message="Passwords do not match." />;
    }
  }

  function resetErrors() {
    setError({ name: false, email: false, password: false });
  }

  return (
    <SafeAreaView style={styles.container}>
      <AppText
        style={{
          alignSelf: "flex-start",
          marginLeft: "5%",
          marginBottom: "5%",
          fontSize: 34,
        }}
        children="Let's get started!"
      />
      <AppTextInput
        style={styles.textInput}
        placeholder="Full name"
        icon="account"
        isError={err.name}
        onChangeText={(name) => {
          setName(name);
          resetErrors();
        }}
      />
      <AppTextInput
        style={styles.textInput}
        placeholder="Email"
        icon="email"
        isError={err.email}
        onChangeText={(email) => {
          setEmail(email);
          resetErrors();
        }}
      />
      <AppTextInput
        style={styles.textInput}
        placeholder="Password"
        icon="lock"
        onChangeText={(password) => {
          setPassword(password);
          resetErrors();
        }}
        isError={err.password}
        secureTextEntry={true}
      />
      <AppTextInput
        style={styles.textInput}
        placeholder="Confirm password"
        icon="lock"
        onChangeText={(password) => {
          setConfirmPassword(password);
          resetErrors();
        }}
        isError={err.password}
        secureTextEntry={true}
      />
      {displayErrorMessage(err)}
      <TextButton style={styles.submitButton} name="Sign Up" onPress={signUp} />
      <Text style={{ marginVertical: "5%" }}>
        Already have an account?
        <Text
          style={{ color: "#0079d3" }}
          onPress={() => {
            console.log("Navigate to the login screen");
          }}
        >
          {" Log in!"}
        </Text>
      </Text>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "white",
    alignItems: "center",
    justifyContent: "center",
  },
  textInput: {
    width: "90%",
  },
  submitButton: { marginVertical: 12, minWidth: "90%" },
});

function ErrorMessage({ message }) {
  return (
    <AppText
      style={{
        alignSelf: "flex-start",
        marginLeft: "6%",
        color: "red",
      }}
      children={message}
    />
  );
}

function validateEmail(email) {
  const re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return re.test(String(email).toLowerCase());
}

function validateFields(fields, setError) {
  const { name, email, password, confirmPassword } = fields;
  const validName = Boolean(name);
  if (!validName) {
    // Only error on the name field
    setError({ name: !validName, email: false, password: false });
    return validName;
  }

  const validEmail = validateEmail(email);
  if (!validEmail) {
    // Only error on the email field
    setError({ name: false, email: !validEmail, password: false });
    return validEmail;
  }

  const validPassword =
    Boolean(password) && password === confirmPassword && password.length >= 8;
  setError({ name: !validName, email: !validEmail, password: !validPassword });
  return validPassword;
}

export default SignupScreen;
