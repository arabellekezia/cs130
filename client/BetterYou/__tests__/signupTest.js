import {render, cleanup, fireEvent} from '@testing-library/react-native';
import React from 'react';
import SignupScreen from "../app/Screens/SignupScreen"

describe('Signup Form', () => {
  let wrapper;
  beforeEach(() => {
      wrapper = render(<SignupScreen />);
    });
  afterEach(() => {
      cleanup();
      wrapper = null;
    });
  it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
  });

  it("Shows invalid email error message when user provides an ill-formatted email", () => {
    const {getByText, queryByTestId, getByPlaceholderText} = wrapper;

    const fullNameInput = getByPlaceholderText("Full name"); 
    const enteredFullName = "Evan Lin";
    fireEvent(fullNameInput, 'onChangeText', enteredFullName);
    
    const emailTextInput = getByPlaceholderText("Email"); 
    const invalidEnteredEmail = "test";
    fireEvent(emailTextInput, 'onChangeText', invalidEnteredEmail);
    
    const signupButton = getByText("Sign Up");
    fireEvent.press(signupButton);
  
    const validationError = queryByTestId("error-message");
    expect(validationError).toBeTruthy();
    
    expect(validationError.props.children)
        .toBe("Please use a valid email.");
  });

  it("Shows missing name error when user doesn't provides their name", () => {
    const {getByText, queryByTestId, getByPlaceholderText} = wrapper;
    
    const signupButton = getByText("Sign Up");
    fireEvent.press(signupButton);
  
    const validationError = queryByTestId("error-message");
    expect(validationError).toBeTruthy();
    
    expect(validationError.props.children)
        .toBe("Please enter your full name.");
  });

  it("Shows password validation error message when the password is less than 8 characters long.", () => {
    const {getByText, queryByTestId, getByPlaceholderText} = wrapper;
    const fullNameInput = getByPlaceholderText("Full name"); 
    const enteredFullName = "Evan Lin";
    fireEvent(fullNameInput, 'onChangeText', enteredFullName);
    
    const emailTextInput = getByPlaceholderText("Email"); 
    const validEnteredEmail = "test@gmail.com";
    fireEvent(emailTextInput, 'onChangeText', validEnteredEmail);

    const passwordTextInput = getByPlaceholderText("Password"); 
    const passwordEntered = "1234567";
    fireEvent(passwordTextInput, 'onChangeText', passwordEntered);

    const signupButton = getByText("Sign Up");
    fireEvent.press(signupButton);
  
    const validationError = queryByTestId("error-message");
    expect(validationError).toBeTruthy();
    
    expect(validationError.props.children)
        .toBe("Password must be at least 8 characters long!");
  });

  it("Shows error when the password and confirm password inputs are mismatched.", () => {
    const {getByText, queryByTestId, getByPlaceholderText} = wrapper;
    const fullNameInput = getByPlaceholderText("Full name"); 
    const enteredFullName = "Evan Lin";
    fireEvent(fullNameInput, 'onChangeText', enteredFullName);
    
    const emailTextInput = getByPlaceholderText("Email"); 
    const validEnteredEmail = "test@gmail.com";
    fireEvent(emailTextInput, 'onChangeText', validEnteredEmail);

    const passwordTextInput = getByPlaceholderText("Password"); 
    const passwordEntered = "12345678";
    fireEvent(passwordTextInput, 'onChangeText', passwordEntered);

    const confirmPasswordTextInput = getByPlaceholderText("Confirm password"); 
    const confirmPasswordEntered = "12345677";
    fireEvent(confirmPasswordTextInput, 'onChangeText', confirmPasswordEntered);

    const signupButton = getByText("Sign Up");
    fireEvent.press(signupButton);
  
    const validationError = queryByTestId("error-message");
    expect(validationError).toBeTruthy();
    
    expect(validationError.props.children)
        .toBe("Passwords do not match.");
  });

  it("No error should be displayed when all valid inputs.", () => {
    const {getByText, queryByTestId, getByPlaceholderText} = wrapper;
    const fullNameInput = getByPlaceholderText("Full name"); 
    const enteredFullName = "Evan Lin";
    fireEvent(fullNameInput, 'onChangeText', enteredFullName);
    
    const emailTextInput = getByPlaceholderText("Email"); 
    const validEnteredEmail = "test@gmail.com";
    fireEvent(emailTextInput, 'onChangeText', validEnteredEmail);

    const passwordTextInput = getByPlaceholderText("Password"); 
    const passwordEntered = "12345678";
    fireEvent(passwordTextInput, 'onChangeText', passwordEntered);

    const confirmPasswordTextInput = getByPlaceholderText("Confirm password"); 
    const confirmPasswordEntered = "12345678";
    fireEvent(confirmPasswordTextInput, 'onChangeText', confirmPasswordEntered);

    const signupButton = getByText("Sign Up");
    fireEvent.press(signupButton);
  
    const validationError = queryByTestId("error-message");
    expect(validationError).toBe(null);
  });
});

