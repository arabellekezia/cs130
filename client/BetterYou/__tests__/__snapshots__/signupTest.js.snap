// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup Form renders correctly 1`] = `
<RCTScrollView
  collapsable={false}
  contentContainerStyle={
    Object {
      "flex": 1,
    }
  }
  keyboardShouldPersistTaps="handled"
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
>
  <View>
    <RCTSafeAreaView
      emulateUnlessSupported={true}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Array [
            Object {},
            Object {
              "alignSelf": "flex-start",
              "fontSize": 34,
              "marginBottom": "5%",
              "marginLeft": "5%",
            },
          ]
        }
      >
        Let's get started!
      </Text>
      <View
        style={
          Object {
            "backgroundColor": "#efefef",
            "borderRadius": 10,
            "flexDirection": "row",
            "marginVertical": 10,
            "padding": 15,
            "width": "90%",
          }
        }
      >
        <Text />
        <TextInput
          allowFontScaling={true}
          maxLength={25}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Full name"
          rejectResponderTermination={true}
          style={
            Object {
              "fontSize": 18,
            }
          }
          underlineColorAndroid="transparent"
        />
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#efefef",
            "borderRadius": 10,
            "flexDirection": "row",
            "marginVertical": 10,
            "padding": 15,
            "width": "90%",
          }
        }
      >
        <Text />
        <TextInput
          allowFontScaling={true}
          autoCapitalize="none"
          keyboardType="email-address"
          maxLength={25}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Email"
          rejectResponderTermination={true}
          style={
            Object {
              "fontSize": 18,
            }
          }
          underlineColorAndroid="transparent"
        />
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#efefef",
            "borderRadius": 10,
            "flexDirection": "row",
            "marginVertical": 10,
            "padding": 15,
            "width": "90%",
          }
        }
      >
        <Text />
        <TextInput
          allowFontScaling={true}
          maxLength={25}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Password"
          rejectResponderTermination={true}
          secureTextEntry={true}
          style={
            Object {
              "fontSize": 18,
            }
          }
          underlineColorAndroid="transparent"
        />
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#efefef",
            "borderRadius": 10,
            "flexDirection": "row",
            "marginVertical": 10,
            "padding": 15,
            "width": "90%",
          }
        }
      >
        <Text />
        <TextInput
          allowFontScaling={true}
          maxLength={25}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Confirm password"
          rejectResponderTermination={true}
          secureTextEntry={true}
          style={
            Object {
              "fontSize": 18,
            }
          }
          underlineColorAndroid="transparent"
        />
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#007AFF",
              "borderRadius": 10,
              "justifyContent": "center",
              "marginVertical": 12,
              "minWidth": "90%",
              "paddingHorizontal": 30,
              "paddingVertical": 12,
            }
          }
        >
          <Text
            style={
              Array [
                Object {},
                Object {
                  "color": "white",
                  "fontSize": 18,
                  "fontWeight": "600",
                },
              ]
            }
          >
            Sign Up
          </Text>
        </View>
      </View>
      <Text
        style={
          Object {
            "marginVertical": "5%",
          }
        }
      >
        Already have an account?
        <Text
          onPress={[Function]}
          style={
            Object {
              "color": "#0079d3",
            }
          }
        >
           Log in!
        </Text>
      </Text>
    </RCTSafeAreaView>
  </View>
</RCTScrollView>
`;
